promql/engine.go

```diff
@@ -1122,6 +1122,13 @@ func (ev *evaluator) rangeEval(prepSeries func(labels.Labels, *EvalSeriesHelper)
 			warnings.Merge(ws)
 			matrixes[i] = val.(Matrix)
 
+			// Since we are copying the samples in memory here, check for ErrTooManySamples.
+			rangeEvalSamples += matrixes[i].TotalSamples()
+			if rangeEvalSamples+ev.currentSamples > ev.maxSamples {
```

Comment:
note for reviewers: we check for `ErrTooManySamples` after evaluating each expression and fail early if `maxSamples` breached.